### PR TITLE
[SPARK-42527][CONNECT] Scala Client add Window functions

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR aims add the window functions to the Scala spark connect client.

### Why are the changes needed?
Provide same APIs in the Scala spark connect client as in the original Dataset API.

### Does this PR introduce _any_ user-facing change?
Yes, it adds new for functions to the Spark Connect Scala client.

### How was this patch tested?

- Add new test
- Manual checked connect-client-jvm and connect with Scala-2.13

### CODE REVIEW

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -129,7 +132,7 @@ object functions {
       case v: Array[Byte] => createLiteral(_.setBinary(ByteString.copyFrom(v)))
       case v: collection.mutable.WrappedArray[_] => lit(v.array)
       case v: LocalDate => createLiteral(_.setDate(v.toEpochDay.toInt))
-      case null => unsupported("Null literals not supported yet.")
+      case null => createLiteral(_.setNull(nullType))
```
Comment (LuciferYang) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:135]: `lag` and `lead` need pass null

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 2]
```diff
@@ -517,6 +517,10 @@ case class Lead(
     input: Expression, offset: Expression, default: Expression, ignoreNulls: Boolean)
     extends FrameLessOffsetWindowFunction with TernaryLike[Expression] {
 
+  def this(
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:520]: without this change, `function_lead` in `ProtoToParsedPlanTestSuite` will failed as follows:

```
- function_lead *** FAILED ***
  org.apache.spark.sql.AnalysisException: [WRONG_NUM_ARGS.WITHOUT_SUGGESTION] The `lead` requires [0, 1, 2, 3] parameters but the actual number is 4. Please, refer to 'https://spark.apache.org/docs/latest/sql-ref-functions.html' for a fix.
  at org.apache.spark.sql.errors.QueryCompilationErrors$.wrongNumArgsError(QueryCompilationErrors.scala:681)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistryBase$.$anonfun$build$7(FunctionRegistry.scala:143)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistryBase$.$anonfun$build$4(FunctionRegistry.scala:139)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistry$.$anonfun$expression$1(FunctionRegistry.scala:875)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistryBase.lookupFunction(FunctionRegistry.scala:233)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistryBase.lookupFunction$(FunctionRegistry.scala:227)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistry.lookupFunction(FunctionRegistry.scala:299)
  at org.apache.spark.sql.catalyst.catalog.SessionCatalog.$anonfun$resolveBuiltinOrTempFunctionInternal$1(SessionCatalog.scala:1644)
  at org.apache.spark.sql.catalyst.catalog.SessionCatalog.lookupTempFuncWithViewContext(SessionCatalog.scala:1654)
  ...

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 3]
```diff
@@ -571,6 +575,10 @@ case class Lag(
     input: Expression, inputOffset: Expression, default: Expression, ignoreNulls: Boolean)
     extends FrameLessOffsetWindowFunction with TernaryLike[Expression] {
 
+  def this(
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala:578]: without this change, `function_lag` in `ProtoToParsedPlanTestSuite` will failed as follows:

```
- function_lag *** FAILED ***
  org.apache.spark.sql.AnalysisException: [WRONG_NUM_ARGS.WITHOUT_SUGGESTION] The `lag` requires [0, 1, 2, 3] parameters but the actual number is 4. Please, refer to 'https://spark.apache.org/docs/latest/sql-ref-functions.html' for a fix.
  at org.apache.spark.sql.errors.QueryCompilationErrors$.wrongNumArgsError(QueryCompilationErrors.scala:681)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistryBase$.$anonfun$build$7(FunctionRegistry.scala:143)
  at scala.Option.getOrElse(Option.scala:189)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistryBase$.$anonfun$build$4(FunctionRegistry.scala:139)
  at org.apache.spark.sql.catalyst.analysis.FunctionRegistry$.$anonfun$expression$1(FunctionRegistry.scala:875)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistryBase.lookupFunction(FunctionRegistry.scala:233)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistryBase.lookupFunction$(FunctionRegistry.scala:227)
  at org.apache.spark.sql.catalyst.analysis.SimpleFunctionRegistry.lookupFunction(FunctionRegistry.scala:299)
  at org.apache.spark.sql.catalyst.catalog.SessionCatalog.$anonfun$resolveBuiltinOrTempFunctionInternal$1(SessionCatalog.scala:1644)
  at org.apache.spark.sql.catalyst.catalog.SessionCatalog.lookupTempFuncWithViewContext(SessionCatalog.scala:1654)
  ...


```

Comment (LuciferYang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: change to fix this in `connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala`

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by hvanhovell** [comment]
Merging this to master/3.4. Thanks for doing this!

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala (hunk 1 above)
Comment (hvanhovell) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:135]: @LuciferYang can you do me a favor and expand this to support all/most what Literal currently does?

Comment (LuciferYang) [reply connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:135]: ok, let me do this later

**Comment by LuciferYang** [comment]
Thanks @hvanhovell

**Pull request closed without merging**